### DISCUSSION TITLE
Error: no decode delegate for this image format `JP2' (JPEG2000)

### DISCUSSION THREAD

**Discussion by pjotrek-b** [ImageMagick/ImageMagick#3427] [General]
Hi everyone!

Summary: This is a format-support feature request.

I've received a set of JPEG2000 image files from a library, but ImageMagick seems to have troubles identifying/opening them. I get the following error when using IM tools on these files:
> identify-im6.q16: no decode delegate for this image format `JP2' @ error/constitute.c/ReadImage/504.

**IM version: 8:6.9.7.4+dfsg-16ubuntu6.9**

Due to content rights, I'm unable to upload the file here directly. I've provided a screenshot of the hexview of the file's header, to maybe give a clue which "flavor" of JP2 this could be. I can provide sample files, but only directly, not publicly.

Is there any chance of adding support for these files to ImageMagick?
(Wouldn't have to be done gratis, btw.)

Thank you very much in advance! :smile: 

![JP2-Header_Hexedit](https://user-images.githubusercontent.com/2419195/111966187-eac06f80-8af6-11eb-9b55-db09d62946dd.png)

**Comment by pjotrek-b** [comment]
Additional information:

MediaInfo output:

> General
> Complete name                            : test.jp2
> Format                                   : JPEG 2000
> Format profile                           : MPEG-4
> Codec ID                                 : jp2  (jp2 )
> File size                                : 7.11 MiB
> 
> Image
> Format                                   : JPEG 2000
> Format profile                           : No restrictions
> Width                                    : 2 180 pixels
> Height                                   : 3 178 pixels
> Chroma subsampling                       : 4:4:4
> Bit depth                                : 8 bits
> Compression mode                         : Lossless

FFmpeg identifies it as:

> Input #0, j2k_pipe, from 'test.jp2':
>   Duration: N/A, bitrate: N/A
>   Stream #0:0: Video: jpeg2000, rgb24, 2180x3178

**Reply by dlemstra** [reply]
Your ImageMagick version is ancient. Have you tried upgrading to a more recent version?

**Reply by pjotrek-b** [reply]
@dlemstra Thanks for pointing that out!
(It didn't feel that long ago that I've setup the server on Ubuntu 18.04 :wink:)

However: I've now tried with "**ImageMagick 7.0.11-4 Q16 x86_64 2021-03-13**": Same.
> identify: no decode delegate for this image format `JP2' @ error/constitute.c/ReadImage/572.

**Reply by dlemstra** [reply]
What is the output of `magick -list version` and what is shown for `JP2` in `magick -list format`?

**Reply by pjotrek-b** [reply]
@dlemstra 
```
$ ./imagemagick-7.0.11 -list version
magick: unrecognized list type '-list' 'version' Initializing @ error/operation.c/CLINoImageOperator/5153.
```

And the other one returns nothing:
`$ ./imagemagick-7.0.11 -list format | grep -i JP2`

:question: :grinning:

**Reply by dlemstra** [reply]
It looks like you don't have `JP2` support built in to ImageMagick and I am sorry about the typo earlier. It should be `-version` instead.

**Reply by pjotrek-b** [reply]
Typo, yeah sorry! Saw it too late.
Nevermind :smile: - see below replies.

**Comment by snibgo** [comment]
@pjotrek-b:

Please show the exact IM command(s) you tried.

Please also show the text output from "exiftool test.jp2".

**Reply by pjotrek-b** [reply]
Here's the output from `exiftool test.jp2`:

[test.jp2-exif.txt](https://github.com/ImageMagick/ImageMagick/files/6185485/test.jp2-exif.txt)

The exact commands I've tried so far were:
```
$ ./imagemagick-7.0.11 identify test.jp2 
identify: no decode delegate for this image format `JP2' @ error/constitute.c/ReadImage/572.
```

```
$ ./imagemagick-7.0.11 convert test.jp2 test.png
convert: no decode delegate for this image format `JP2' @ error/constitute.c/ReadImage/572.
convert: no images defined `test.png' @ error/convert.c/ConvertImageCommand/3322.
```

Nothing else yet, since it looks to me like IM can't understand that JP2-flavor *yet* :grin:
If that's the case, wouldn't it be awesome if it'd be possible to get a patch upstream to add support for that?

Anyone interested?
Please contact me! :+1:

**Comment by fmw42** [comment]
Does `magick -version` list jp2?  Do you have OpenJPEG library 2.1 or higher installed? IM no longer uses the Jasper delegate and uses OpenJPEG, since it is more functional.

**Reply by pjotrek-b** [reply]
Good question... How do I check?

`magick -version` returns this:
```
Version: ImageMagick 7.0.11-4 Q16 x86_64 2021-03-13 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(3.1) 
Delegates (built-in): bzlib djvu fontconfig freetype gvc jbig jng jpeg lcms lqr lzma openexr png raqm tiff x xml zlib
```

I have installed `libopenjp2-7` and `libopenjpip7`.
It's the [AppImage](https://download.imagemagick.org/ImageMagick/download/binaries/magick) downloaded from the [official download page](https://imagemagick.org/script/download.php#linux).

**Reply by pjotrek-b** [reply]
Strange:
If I `./imagemagick-7.0.11 convert DH_object_foto.jpg -quality 0 delme.jp2`, it creates the file without any errors or warnings - but the "delme.jp2" is a JPEG (not 2000). According to `identify`and `mediainfo`.

How do I enable JPEG2000 support in IM?
[Awesome IM devs seem to have already implemented some...](https://imagemagick.org/script/jp2.php)?

Thanks for any help! :smile:

**Comment by snibgo** [comment]
What does "convert -version" say?

Why are you using "convert", "identify" etc?

**Reply by pjotrek-b** [reply]
The `convert` and `identify`commands are the ones coming from the distro repositories, so you can ignore that, as it's being superseded by me using the most recent AppImage.

The issue seems to be that the binaries I'm using seem to be missing JP2 support.
I found a related thread on AskUbuntu from 2017:
[How to enable JPEG2000 (jp2) in ImageMagick?](https://askubuntu.com/questions/899541/how-to-enable-jpeg2000-jp2-in-imagemagick)

I can try to compile IM, but I'm surprised it's necessary since it looks like the JP2 dependencies are in the default repositories?

What is the proper/regular way of using IM with JPEG2000 (on Linux)?


Thanks again :smile:
